### PR TITLE
REF: make array_to_datetime single-pass

### PR DESCRIPTION
**Pull request by jbrockmendel**
Or more specifically, remove the need for falling back to datetime_to_datetime64 which does a second pass.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
cc @mroeschke

File: pandas/_libs/tslib.pyx [hunk 1]
```diff
@@ -457,26 +461,35 @@ cpdef array_to_datetime(
...
-                            _ts = convert_datetime_to_tsobject(val, None)
-                            iresult[i] = _ts.value
-                        else:
+                        seen_aware = True
+
+                        if seen_tz is None:
+                            seen_tz = val.tzinfo
+                            if not utc_convert:
+                                tz_out = val.tzinfo
+                        elif not utc_convert and not tz_compare(seen_tz, val.tzinfo):
```
Comment (jreback) [inline pandas/_libs/tslib.pyx:470]: is this comparison expenseive?

**Review by jreback — CHANGES_REQUESTED** [review]
how do asv's look here?

File: pandas/_libs/tslib.pyx [hunk 2]
```diff
@@ -457,26 +461,35 @@ cpdef array_to_datetime(
                 elif PyDateTime_Check(val):
                     seen_datetime = True
                     if val.tzinfo is not None:
-                        if utc_convert:
-                            _ts = convert_datetime_to_tsobject(val, None)
-                            iresult[i] = _ts.value
-                        else:
+                        seen_aware = True
+
+                        if seen_tz is None:
```
Comment (mroeschke) [inline pandas/_libs/tslib.pyx:466]: What happens when we get a timezone offset from a string and a tzinfo object from a `datetime.datetime`?

**Comment by jreback** [comment]
can you rebase

**Comment by jbrockmendel** [comment]
> can you rebase

yah will circle back before too long, need to address some other comments above.  pretty focused on MultiIndex bugs ATM.

**Comment by jbrockmendel** [comment]
mothballing to clear the queue, will revisit

**Pull request closed without merging**